docs/rfcs/041-rel-sparse-keyspace.md

```diff
@@ -0,0 +1,176 @@
+information needs to be preserved from the parent branch to the child branch. Therefore, the read path needs
+to be updated accordingly to accommodate such "inherited sparse keys". This is done in
+[PR#10313](https://github.com/neondatabase/neon/pull/10313).
+
+## Coexistence of the Old and New Keyspaces
+
+Migrating to the new keyspace will be done gradually: when we flip a config item to enable the new reldir keyspace, the
+ingestion path will start to write to the new keyspace and the old reldir data will be kept in the old one. The read
+path needs to combine the data from both keyspaces.
+
+Theoretically, we could do a rewrite at the startup time that scans all relation directories and copies that data into the
+new keyspace. However, this could take a long time, especially if we have thousands of tenants doing the migration
+process simultaneously after the pageserver restarts. Therefore, we propose the coexistence strategy so that the
+migration can happen seamlessly and imposes no potential downtime for the user.
+
+With the coexistence assumption, the 3 reldir operations will be implemented as follows:
+
+1. Check if a relation exists
+  - Check the new keyspace if the key maps to any value. If it maps to "exists" or "deleted", directly
+    return it to the user.
+  - Otherwise, deserialize the old reldir key and get the result.
+2. List all relations: scan the sparse keyspace over the `rel_dir_key_prefix` and deserialize the old reldir key.
+   Combine them to obtain the final result.
+3. Create/drop a relation: write "exists" or "deleted" to the corresponding key of the relation into the new keyspace.
+  - Alternatively, when a user removes a relation when the new keyspace is enabled but it is created in the
+    old reldir key, we can remove it from the old reldir key and write the delete tombstone in the new keyspace.
+
+We will introduce a config item and an index_part record to record the current status of the migration process.
+
+* Config item `enable_reldir_v2`: controls whether the ingestion path writes the reldir info into the new keyspace.
+* `index_part.json` field `reldir_v2_status`: whether the timeline has written any key into the new reldir keyspace.
+
+If `enable_reldir_v2` is set to `true` and the timeline ingests the first key into the new reldir keyspace, it will update
+`index_part.json` to set `reldir_v2_status` to `Status::Migrating`. Even if `enable_reldir_v2` gets flipped back to
+`false` (i.e., when the pageserver restarts and such config isn't persisted), the read/write path will still
+read/write to the new keyspace to avoid data inconsistency.
+
+## Full Migration
+
+This won't be implemented in the project's first phase but might be implemented in the future. Having both v1 and
+v2 existing in the system would force us to keep the code to deserialize the old reldir key forever. To entirely deprecate this
+code path, we must ensure the timeline has no old reldir data.
```

Comment:
> Having both v1 and v2 existing in the system would force us to keep the code to deserialize the old reldir key forever.

I think we'll end up supporting both for a very long time with this approach. We detach inactive tenants from the pageserver, so we depend on compute activity to perform the full migration.

Reply:
I'm fine with having two read paths and I feel this is the best way to seamlessly migrate all customers. The ingestion path will only modify the v2 reldirs after we finish this project.

Reply:
I agree that the read-through approach makes sense, but I think we should eagerly perform a full migration in the near term, even if that means briefly re-attaching all tenants (can we make that a light-weight operation?).

Reply:
It's hard to do that until we do a full migration to the storcon. We need to go through cplane to get a list of all tenants and attach tenants safely without getting multi-attached. I don't think we've ever tried re-attaching all tenants before...

Reply:
Waiting for storcon is probably fine, I think that should happen within weeks.